src/EditorFeatures/TestUtilities/Workspaces/TestWorkspace.cs

```diff
@@ -704,5 +707,52 @@ internal ITextBuffer GetOrCreateBufferForPath(string? filePath, IContentType con
                 return textBuffer;
             });
         }
+        internal override bool TryApplyChanges(
+            Solution newSolution,
+            IProgressTracker progressTracker)
+        {
+            var oldSolution = CurrentSolution;
+            if (base.TryApplyChanges(newSolution, progressTracker))
+            {
+                var projectChanges = newSolution.GetChanges(oldSolution).GetProjectChanges().ToList();
+                var changedDocs = projectChanges.SelectMany(pd => pd.GetChangedDocuments(true).Concat(pd.GetChangedAdditionalDocuments())).ToList();
+
+                NotifyRefactorChanges(changedDocs, newSolution, oldSolution);
+                return true;
+            }
+
+            return false;
+        }
+
+        private void NotifyRefactorChanges(
+            IEnumerable<DocumentId> changedDocumentIds,
+            Solution newSolution,
+            Solution oldSolution,
+            CancellationToken cancellationToken = default)
+        {
+            // Without any services to notify there's no need to dig through the documents
+            if (!_refactorNotifyServices.Any())
+            {
+                return;
+            }
+
+            var changedSymbols = Rename.RenameSymbolAnnotation
+                .GatherChangedSymbolsInDocumentsAsync(changedDocumentIds, newSolution, oldSolution, cancellationToken)
+                .WaitAndGetResult_CanCallOnBackground(cancellationToken);
```

Comment:
we're waiting on the UI thread? (since TryApplyChanges should only be on the UI thread)?  this seems bad.  could'nt this hang for the user indefinitely without any clue what's going on?

Reply:
Without waiting, would continuation be the best? Since this is working on immutable solutions it should be fine to capture and have "eventual correctness". If I change to a synchronous method, it will need to call synchronous APIs that could also block for semantic model and syntax root right?

Reply:
probably best to discuss this with jason.  However, if you're calling for semanticmodels/syntax/etc. it sounds likke you're basically unbounded in cost.  So we need a way to safely do that stuff in these codepaths in a reasonable way for the user.

That either means making this all async, or making sure there's a dialog the user can interact with, or maybe something else.  Right now, it looks like you could basically just hang the VS ui perpetually with the user unable to do anything about it :(